File: lib/analyzer/sections.ts
// Section headers are often ALL CAPS, Title Case, or followed by a colon.
// We match at the start of a line and allow an optional colon/dash suffix.
const SECTION_PATTERNS: Array<{ key: string; regex: RegExp }> = [
  { key: 'summary',          regex: /^(summary|profile|objective|about|professional summary|career objective)\s*:?$/im },
  { key: 'experience',       regex: /^(experience|work experience|employment|professional experience|work history|career history)\s*:?$/im },
  { key: 'education',        regex: /^(education|academic background|academic history|degrees?|schooling|qualifications)\s*:?$/im },
  { key: 'skills',           regex: /^(skills|technical skills|core competencies|technologies|expertise|proficiencies)\s*:?$/im },
  { key: 'projects',         regex: /^(projects|project experience|portfolio|personal projects|side projects|open.?source)\s*:?$/im },
  { key: 'certifications',   regex: /^(certifications?|licenses?|credentials|accreditations?|professional development)\s*:?$/im },
  { key: 'awards',           regex: /^(awards?|honors?|achievements?|recognition)\s*:?$/im },
  { key: 'publications',     regex: /^(publications?|research|papers?|articles?)\s*:?$/im },
]

export function detectSections(text: string): Record<string, string> {
  const lines = text.split('\n')
  const sections: Record<string, string> = {}
  let currentKey = 'other'
  const buffers: Record<string, string[]> = { other: [] }

  for (const line of lines) {
    const trimmed = line.trim()
    const match = SECTION_PATTERNS.find(({ regex }) => regex.test(trimmed))

    if (match) {
      currentKey = match.key
      buffers[currentKey] ??= []
    } else {
      buffers[currentKey] ??= []
      buffers[currentKey].push(line)
    }
  }

  for (const [key, lines] of Object.entries(buffers)) {
    const content = lines.join('\n').trim()
    if (content) sections[key] = content
  }

  return sections
}

File: api/app/services/sections.py
import re

# Patterns mirror the TypeScript sections.ts detector so both engines
# produce consistent section keys (experience, skills, education, etc.).
SECTION_PATTERNS: list[tuple[str, re.Pattern]] = [
    ("summary",        re.compile(r"^(summary|profile|objective|about|professional summary|career objective)\s*:?$", re.I | re.M)),
    ("experience",     re.compile(r"^(experience|work experience|employment|professional experience|work history|career history)\s*:?$", re.I | re.M)),
    ("education",      re.compile(r"^(education|academic background|academic history|degrees?|schooling|qualifications)\s*:?$", re.I | re.M)),
    ("skills",         re.compile(r"^(skills|technical skills|core competencies|technologies|expertise|proficiencies)\s*:?$", re.I | re.M)),
    ("projects",       re.compile(r"^(projects|project experience|portfolio|personal projects|side projects|open.?source)\s*:?$", re.I | re.M)),
    ("certifications", re.compile(r"^(certifications?|licenses?|credentials|accreditations?|professional development)\s*:?$", re.I | re.M)),
    ("awards",         re.compile(r"^(awards?|honors?|achievements?|recognition)\s*:?$", re.I | re.M)),
    ("publications",   re.compile(r"^(publications?|research|papers?|articles?)\s*:?$", re.I | re.M)),
]


def detect_sections(text: str) -> dict[str, str]:
    lines = text.splitlines()
    buffers: dict[str, list[str]] = {"other": []}
    current = "other"

    for line in lines:
        trimmed = line.strip()
        matched_key = None
        for key, pattern in SECTION_PATTERNS:
            if pattern.match(trimmed):
                matched_key = key
                break

        if matched_key:
            current = matched_key
            buffers.setdefault(current, [])
        else:
            buffers.setdefault(current, [])
            buffers[current].append(line)

    return {k: "\n".join(v).strip() for k, v in buffers.items() if "\n".join(v).strip()}
